### ISSUE TITLE
podman generate kube ignore "_" in container name.

### ISSUE THREAD

**Issue by kaiyasuda** [containers/podman#7020]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When podman generate kube writes container name, ignore "_".

**Steps to reproduce the issue:**

1.  check running container include "_" in its name.

```
# podman container ls -a
CONTAINER ID  IMAGE                              COMMAND     CREATED        STATUS            PORTS                   NAMES
128d10cc7413  docker.io/library/postgres:latest  postgres    6 minutes ago  Up 6 minutes ago  0.0.0.0:8080->8080/tcp  service_postgres_1
14dfcf8959ac  k8s.gcr.io/pause:3.2                           6 minutes ago  Up 6 minutes ago  0.0.0.0:8080->8080/tcp  84d68cb240c6-infra
945fedadb3e8  quay.io/keycloak/keycloak:latest   -b 0.0.0.0  6 minutes ago  Up 6 minutes ago  0.0.0.0:8080->8080/tcp  service_keycloak_1
```

2. make kubernetes yaml with `generate kube`

```
# podman generate kube service > kubernetes-keycloak-postgres.yml
```

**Describe the results you received:**

3.  check containers name.

```
# cat kubernetes-keycloak-postgres.yml | grep "name: service."
    name: servicekeycloak1
    name: servicepostgres1
```

**Describe the results you expected:**

```
# cat kubernetes-keycloak-postgres.yml | grep "name: service."
    name: service_keycloak_1
    name: service_postgres_1
```

**Output of `podman version`:**

```
# podman version
Version:      2.0.2
API Version:  1
Go Version:   go1.13.4
Built:        Wed Dec 31 19:00:00 1969
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
# podman info --debug
host:
  arch: amd64
  buildahVersion: 1.15.0
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.19-1.el8.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.19, commit: 4726cba6f219c7479c28f0687868bd2ffe894869'
  cpus: 2
  distribution:
    distribution: '"centos"'
    version: "8"
  eventLogger: file
  hostname: ycentos8.example.com
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 4.18.0-147.el8.x86_64
  linkmode: dynamic
  memFree: 5390655488
  memTotal: 8191717376
  ociRuntime:
    name: runc
    package: runc-1.0.0-60.rc8.module_el8.1.0+237+63e26edc.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.1-dev'
  os: linux
  remoteSocket:
    path: /run/podman/podman.sock
  rootless: false
  slirp4netns:
    executable: ""
    package: ""
    version: ""
  swapFree: 6874460160
  swapTotal: 6874460160
  uptime: 123h 36m 21.22s (Approximately 5.12 days)
registries:
  search:
  - registry.redhat.io
  - registry.access.redhat.com
  - quay.io
  - docker.io
store:
  configFile: /etc/containers/storage.conf
  containerStore:
    number: 3
    paused: 0
    running: 2
    stopped: 1
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /var/lib/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 11
  runRoot: /var/run/containers/storage
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 1
  Built: 0
  BuiltTime: Wed Dec 31 19:00:00 1969
  GitCommit: ""
  GoVersion: go1.13.4
  OsArch: linux/amd64
  Version: 2.0.2
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
# rpm -q podman
podman-2.0.2-2.el8.x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
VMware

**Comment by zhangguanzhang** [comment]
could you put the file content to there which `podman play kube xxx.yml`?

**Comment by kaiyasuda** [comment]
> could you put the file content to there which `podman play kube xxx.yml`?

this pod made by podman-compose.
compose yml file is here.
https://github.com/keycloak/keycloak-containers/blob/master/docker-compose-examples/keycloak-postgres.yml

```keycloak-postgres.yml
version: '3'

volumes:
  postgres_data:
      driver: local

services:
  postgres:
      image: postgres
      volumes:
        - postgres_data:/var/lib/postgresql/data
      environment:
        POSTGRES_DB: keycloak
        POSTGRES_USER: keycloak
        POSTGRES_PASSWORD: password
  keycloak:
      image: quay.io/keycloak/keycloak:latest
      environment:
        DB_VENDOR: POSTGRES
        DB_ADDR: postgres
        DB_DATABASE: keycloak
        DB_USER: keycloak
        DB_SCHEMA: public
        DB_PASSWORD: password
        KEYCLOAK_USER: admin
        KEYCLOAK_PASSWORD: Pa55w0rd
        # Uncomment the line below if you want to specify JDBC parameters. The parameter below is just an example, and it shouldn't be used in production without knowledge. It is highly recommended that you read the PostgreSQL JDBC driver documentation in order to use it.
        #JDBC_PARAMS: "ssl=true"
      ports:
        - 8080:8080
      depends_on:
        - postgres
```

and this is full list of reconverted to kubernetes yaml file.

```kubernetes-keycloak-postgres.yml
# Generation of Kubernetes YAML is still under development!
#
# Save the output of this file and use kubectl create -f to import
# it into Kubernetes.
#
# Created with podman-2.0.2
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2020-07-20T06:47:49Z"
  labels:
    app: service
  name: service
spec:
  containers:
  - command:
    - docker-entrypoint.sh
    - postgres
    env:
    - name: PATH
      value: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/lib/postgresql/12/bin
    - name: TERM
      value: xterm
    - name: HOSTNAME
    - name: POSTGRES_USER
      value: keycloak
    - name: DB_PASSWORD
      value: password
    - name: JDBC_MYSQL_VERSION
      value: 8.0.19
    - name: JBOSS_HOME
      value: /opt/jboss/keycloak
    - name: JDBC_MSSQL_VERSION
      value: 7.4.1.jre11
    - name: KEYCLOAK_USER
      value: admin
    - name: DB_VENDOR
      value: POSTGRES
    - name: container
      value: podman
    - name: DB_SCHEMA
      value: public
    - name: PROXY_ADDRESS_FORWARDING
      value: "false"
    - name: DB_DATABASE
      value: keycloak
    - name: JDBC_MARIADB_VERSION
      value: 2.5.4
    - name: PG_MAJOR
      value: "12"
    - name: DB_USER
      value: keycloak
    - name: KEYCLOAK_VERSION
      value: 10.0.2
    - name: LANG
      value: en_US.utf8
    - name: PG_VERSION
      value: 12.3-1.pgdg100+1
    - name: DB_ADDR
      value: postgres
    - name: GOSU_VERSION
      value: "1.12"
    - name: LAUNCH_JBOSS_IN_BACKGROUND
      value: "1"
    - name: POSTGRES_PASSWORD
      value: password
    - name: POSTGRES_DB
      value: keycloak
    - name: KEYCLOAK_PASSWORD
      value: Pa55w0rd
    - name: JDBC_POSTGRES_VERSION
      value: 42.2.5
    - name: PGDATA
      value: /var/lib/postgresql/data
    image: docker.io/library/postgres:latest
    name: postgres_1
    ports:
    - containerPort: 8080
      hostPort: 8080
      protocol: TCP
    resources: {}
    securityContext:
      allowPrivilegeEscalation: true
      capabilities: {}
      privileged: false
      readOnlyRootFilesystem: false
      seLinuxOptions: {}
    workingDir: /
  - command:
    - /opt/jboss/tools/docker-entrypoint.sh
    - -b
    - 0.0.0.0
    env:
    - name: PATH
      value: /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
    - name: TERM
      value: xterm
    - name: HOSTNAME
    - name: JBOSS_HOME
      value: /opt/jboss/keycloak
    - name: JDBC_MSSQL_VERSION
      value: 7.4.1.jre11
    - name: LANG
      value: en_US.UTF-8
    - name: KEYCLOAK_USER
      value: admin
    - name: DB_PASSWORD
      value: password
    - name: KEYCLOAK_VERSION
      value: 10.0.2
    - name: JDBC_MYSQL_VERSION
      value: 8.0.19
    - name: DB_SCHEMA
      value: public
    - name: PROXY_ADDRESS_FORWARDING
      value: "false"
    - name: DB_DATABASE
      value: keycloak
    - name: DB_VENDOR
      value: POSTGRES
    - name: container
      value: podman
    - name: DB_ADDR
      value: postgres
    - name: LAUNCH_JBOSS_IN_BACKGROUND
      value: "1"
    - name: JDBC_MARIADB_VERSION
      value: 2.5.4
    - name: KEYCLOAK_PASSWORD
      value: Pa55w0rd
    - name: JDBC_POSTGRES_VERSION
      value: 42.2.5
    - name: DB_USER
      value: keycloak
    image: quay.io/keycloak/keycloak:latest
    name: keycloak_1
    resources: {}
    securityContext:
      allowPrivilegeEscalation: true
      capabilities: {}
      privileged: false
      readOnlyRootFilesystem: false
      runAsGroup: 0
      runAsUser: 1000
      seLinuxOptions: {}
    workingDir: /
status: {}
---
metadata:
  creationTimestamp: null
spec: {}
status:
  loadBalancer: {}
```

I fix container name and play this reconverted file, but it was not solve.

```
# cat kubernetes-keycloak-postgres.yml | grep "    name: ."
    name: postgres_1
    name: keycloak_1
```

```
# podman play kubernetes-keycloak-postgres.yml
# podman container ls -a
CONTAINER ID  IMAGE                              COMMAND               CREATED         STATUS             PORTS                   NAMES
3e759ac0523b  quay.io/keycloak/keycloak:latest   /opt/jboss/tools/...  18 seconds ago  Up 16 seconds ago  0.0.0.0:8080->8080/tcp  service-keycloak_1
7a67cd3ad752  k8s.gcr.io/pause:3.2                                     26 seconds ago  Up 18 seconds ago  0.0.0.0:8080->8080/tcp  4d2a256d3921-infra
d833d321b8ec  docker.io/library/postgres:latest  docker-entrypoint...  22 seconds ago  Up 17 seconds ago  0.0.0.0:8080->8080/tcp  service-postgres_1
```

```
# podman generate service > kubernetes-keycloak-postgres2.yml
# cat kubernetes-keycloak-postgres2.yml | grep "name: service."
    name: service-keycloak1
    name: service-postgres1
```

**Comment by rhatdan** [comment]
@haircommander PTAL

**Comment by haircommander** [comment]
https://github.com/containers/podman/blob/17f9b80600bc008e7c0a4060ff3a6bb5eb56d0cc/libpod/kube.go#L252

git blame points to @baude's original implementation of generate kube
https://github.com/containers/podman/commit/f11a74e7150c2929a0ed18732ee0c4895f7e75be

any reason to keep it? I also see we're scrubbing podName for underscores

**Comment by zhangguanzhang** [comment]
@haircommander  PTAL this https://github.com/containers/podman/pull/7002

**Comment by mheon** [comment]
I think there was a reason for it, but I can't recall what it actually was. @baude might be able to tell us were he not on vacation.

**Comment by mheon** [comment]
I thought it might be because we don't allow `_` in container/pod names, but I checked and we do allow them.

**Comment by rhatdan** [comment]
@baude Any ideas on this one?

**Comment by mheon** [comment]
@ParkerVR PTAL. Since no one remembers why we stripped them originally, I think it should be safe to put them in.

**Comment by mheon** [comment]
Fixed by #7587
